indimail-x/adminclient.8
.\" vim: tw=75
.TH adminclient 8
.SH NAME
adminclient \- Client for running IndiMail programs remotely

.SH SYNOPSYS
\fBadminclient\fR \fB\-h\fR \fIadminHost\fR \fB\-p\fR \fIport\fR
\fB\-u\fR \fIadminUser\fR \fB\-P\fR \fIadminPassword\fR
[\fB\-n\fR \fIcertfile\fR] \fB\-c\fR \fICommand\fR

.SH DESCRIPTION
.PP
\fBadminclient\fR is the administration client for administering IndiMail's
clustered domains. It has a simple mechanism by which administrators can
connect to the administration server, \fBindisrvr\fR and issue commands to
execute any of the IndiMail programs \fICommand\fR.  On successful
validation of password, \fBadminclient\fR executes the specified IndiMail
administration program on the remote host by connecting to
\fBindisrvr(8)\fR. For security reasons, the list of these administration
programs are hard-coded in the \fBadminclient\fR and \fBindisrvr\fR
executable. \fBadminclient\Fr uses environment variable \fBADMIN_TIMEOUT\fR
to timeout connections to \fBindisrvr\fR.  \fBadminclient\fR needs to
supply credentials of users with administration privileges. These users can
be created using the \fBmgmtpass(8)\fR program. You can furhter use
\fBvpriv(8)\fR to control/restrict access of these admin users to certain
programs (and even command line arguments.

.PP
\fBadminclient\fR encrypts the communication channel using TLS if \fB\-n\fR
option specifing a certificate is used. One can use use ciphers by setting
\fBTLS_CIPHER_LIST\fR environment variable for TLS protocol TLSv1.2 and
below and \fBTLS_CIPHER_SUITE\fR environment variable for TLS protocol
TLSv1.3 and above. For multiple ciphers, use ':' separated list of ciphers.
See openssl-ciphers(1ossl).

e.g. The following command runs the command vadduser with arguments 
test@example.com password.

.EX
adminclient -h phoenix2 -p  4000 -u admin -P adminpass -c "vadduser test@example.com password"
.EE

.SH OPTIONS
.TP
\fB\-v\fR
Sets Verbose option.
.TP
\fB\-h\fR \fIadminHost\fR
The IP address / hostname of host on which command \fICommand\fR needs to run. 
.TP
\fB\-p\fR \fIadminPort\fR
A TCP/IP port or service in /etc/services on which indisrvr is running on
adminHost.
.TP
\fB\-u\fR \fIadminUser\fR
The administration user (maintained using \fBmgmtpass(8)\fR) for which
indisrvr will validate the password.
.TP
\fB\-P\fR \fIadminPass\fR
The password for the user specified by the -u option.
.TP
\fB\-n\fR \fIcertificate\fR
A client certificate which should be provided in case TLS encryption is
needed
.TP
\fB\-C\fR \fIcafile\fR
Specify \fIcafile\fR as the file containing CA certificates. Sometimes,
special CA certificates are needed if an intermediate certificate used
by your issuing CA expires. Note that the file given with -c is passed
to SSL_CTX_load_verify_locations and thus may contain several CA
certificates, which will be used during SSL connection negotiation to
close any gap in the certificate chain.
.TP
\fB\-r\fR \fIcrlfile\fR
A list of Certificate Revocation Lists (CRLs). If present it
should contain the CRLs of the CAs in \fIcafile\fR and client certs that
will be checked for revocation.
.TP
\fB\-m\fR
Match \fIhost\fR with common name in certificate (CN)
.TP
\fB\-c\fR \fICommand\fR
The command that needs to executed on the remote host \fIadminHost\fR.
Command should be enclosed in quotes for commands requiring multiple
arguments on the command line.

The -h, -p, -u, -P, -v are optional. If not given, the environment
variables ADMIN_HOST, ADMIN_PORT, ADMIN_USER, ADMIN_PASS will be used for
these variables.

.SH RETURN VALUE
Returns the return status of the remote command executed. All output
printed by the remote command on stdout/stderr are printed on stdout of
adminclient.

.SH PROTOCOL
The adminclient protocol is descibed below. You can connect to port 4000
and execute the below sequence of commands to execute. Here 's' denotes
server and 'c' denotes client.

.EX
.B adminclient Protocol
 s - server
 c - client

 s: "Login: "
 c: "userid\\n"
 s: "Password: "
 c: "password\\n"
 s: "OK\\n"
 c: "index command arg1 arg2 ...\\n"
 s: <output of above command if any>
 c: "\\n"
 s: "RETURNSTATUS[return value of command]\\n"
.EE

The below shows an actual example of vuserinfo getting executed by
executing the \fBtelnet\fR(1) client on port 4000.

.EX
 <lf> - linefeed

 Login: admin<lf>
 Password: xxxxxxxx<lf>
 OK
 7 vuserinfo -n manvendra@indimail.org<lf>
 name          : manvendra@indimail.org
 <lf>
 RETURNSTATUS0
.EE

.SH "SEE ALSO"
indisrvr(8), mgmtpass(8), vpriv(8), IndiMail(7), openssl-ciphers(1ossl)


indimail-x/NEWS
doc/RELEASE-Notes-indimail.md

indimail-x/adminclient.c
/*
 * $Log: adminclient.c,v $
 * Revision 1.5  2023-08-08 10:34:36+05:30  Cprogrammer
 * use strerr_tls for reporting tls error
 *
 * Revision 1.4  2023-01-03 21:05:41+05:30  Cprogrammer
 * added -r option to specify certification revocation list
 *
 * Revision 1.3  2021-03-04 11:51:59+05:30  Cprogrammer
 * added -m option to match host with common name
 * added -C option to specify cafile
 *
 * Revision 1.2  2019-04-22 22:24:12+05:30  Cprogrammer
 * added missing header strerr.h
 *
 * Revision 1.1  2019-04-20 08:10:20+05:30  Cprogrammer
 * Initial revision
 *
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#ifndef lint
static char     sccsid[] = "$Id: adminclient.c,v 1.5 2023-08-08 10:34:36+05:30 Cprogrammer Exp mbhangui $";
#endif

#ifdef CLUSTERED_SITE
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_CTYPE_H
#include <ctype.h>
#endif
#ifdef HAVE_QMAIL
#include <stralloc.h>
#include <strerr.h>
#include <strmsg.h>
#include <fmt.h>
#include <str.h>
#include <env.h>
#include <tls.h>
#endif
#include "auth_admin.h"
#include "adminCmmd.h"
#include "variables.h"

char            strnum[FMT_ULONG];

static void
die_nomem()
{
	strerr_warn1("adminclient: out of memory", 0);
	_exit(111);
}

int
main(int argc, char **argv)
{
	char           *admin_user, *admin_pass, *admin_host, *admin_port, *cmmd,
				   *cmdptr1, *cmdptr2, *certfile, *cafile, *crlfile;
	static stralloc cmdbuf = {0}, cmdName = {0};
	int             sfd, i, j, k, input_read, match_cn = 0;

	certfile = cafile = crlfile = NULL;
	admin_user = admin_pass = admin_host = admin_port = cmmd = NULL;
	input_read = 0;
	for (i = 1; i < argc; i++) {
		if (argv[i][0] != '-')
			continue;
		switch (argv[i][1])
		{
		case 'h':
			admin_host = *(argv + i + 1);
			break;
		case 'p':
			admin_port = *(argv + i + 1);
			break;
		case 'u':
			admin_user = *(argv + i + 1);
			break;
		case 'P':
			admin_pass = *(argv + i + 1);
			break;
		case 'c':
			cmmd = *(argv + i + 1);
			break;
		case 'i':
			input_read = 1;
			break;
		case 'n':
			certfile = *(argv + i + 1);
			break;
		case 'C':
			cafile = *(argv + i + 1);
			break;
		case 'r':
			crlfile = *(argv + i + 1);
			break;
		case 'm':
			match_cn = 1;
			break;
		case 'v':
			verbose = 1;
			break;
		default:
			strerr_warn1("usage: adminclient -h adminHost -p adminPort -u adminUser -P adminPasswd [-n certfile] [-i] -c cmmd [-v]", 0);
			strnum[0] = argv[i][1];
			strnum[1] = 0;
			strerr_warn2("invalid option ", strnum, 0);
			return (1);
		}
	}
	if (!admin_host && !(admin_host = (char *) env_get("ADMIN_HOST"))) {
		strerr_warn1("usage: adminclient -h adminHost -p adminPort -u adminUser -P adminPasswd [-n certfile] [-i] -c cmmd [-v]", 0);
		return (1);
	} else
	if (!admin_port && !(admin_port = (char *) env_get("ADMIN_PORT"))) {
		strerr_warn1("usage: adminclient -h adminHost -p adminPort -u adminUser -P adminPasswd [-n certfile] [-i] -c cmmd [-v]", 0);
		return (1);
	} else
	if (!admin_user && !(admin_user = (char *) env_get("ADMIN_USER"))) {
		strerr_warn1("usage: adminclient -h adminHost -p adminPort -u adminUser -P adminPasswd [-n certfile] [-i] -c cmmd [-v]", 0);
		return (1);
	} else
	if (!admin_pass && !(admin_pass = (char *) env_get("ADMIN_PASS"))) {
		strerr_warn1("usage: adminclient -h adminHost -p adminPort -u adminUser -P adminPasswd [-n certfile] [-i] -c cmmd [-v]", 0);
		return (1);
	} else
	if (!cmmd) {
		strerr_warn1("usage: adminclient -h adminHost -p adminPort -u adminUser -P adminPasswd [-n certfile] [-i] -c cmmd [-v]", 0);
		return (1);
	}
	if (verbose)
		strmsg_out7("connecting to ", admin_host, "@", admin_port, " as ", admin_user, "\n");
	if ((sfd = auth_admin(admin_user, admin_pass, admin_host, admin_port, certfile, cafile, crlfile, match_cn)) == -1) {
		strerr_warn7("adminclient: auth_admin: ", admin_host, "@", admin_port, " as ", admin_user, ": ", &strerr_tls);
		return (1);
	}
	if (verbose)
		strmsg_out1("connected\n");
	for (cmdptr1 = cmmd; *cmdptr1 && isspace((int) *cmdptr1); cmdptr1++);
	cmdptr2 = cmdptr1;
	for (; *cmdptr2 && !isspace((int) *cmdptr2); cmdptr2++);
	if (!stralloc_copyb(&cmdName, cmdptr1, cmdptr2 - cmdptr1) ||
			!stralloc_0(&cmdName))
		die_nomem();
	cmdName.len--;
	i = str_rchr(cmdName.s, '/');
	if (cmdName.s[i])
		cmdptr1 = cmdName.s + i + 1;
	else
		cmdptr1 = cmdName.s;
	for (k = 0; adminCommands[k].name; k++) {
		j = str_rchr(adminCommands[k].name, '/');
		if (adminCommands[k].name[j])
			cmdptr2 = adminCommands[k].name + j + 1;
		else
			cmdptr2 = adminCommands[k].name;
		if (!str_diffn(cmdptr1, cmdptr2, cmdName.len - (i + 2))) {
			if (!stralloc_copyb(&cmdbuf, strnum, fmt_uint(strnum, (unsigned int) k)) ||
					!stralloc_append(&cmdbuf, " ") ||
					!stralloc_cats(&cmdbuf, cmmd) ||
					!stralloc_0(&cmdbuf))
				die_nomem();
			cmdbuf.len--;
			if (verbose)
				strmsg_out5("executing command no ", strnum, " [", cmmd, "]\n");
			return (adminCmmd(sfd, input_read, cmdbuf.s, cmdbuf.len));
		}
	}
	strerr_warn2("adminclient: invalid or unauthorized command ", cmmd, 0);
	return (1);
}
#else
#include <strerr.h>
int
main()
{
	strerr_warn1("IndiMail not configured with --enable-user-cluster=y", 0);
	return (1);
}
#endif
